### PR TITLE
Making changelog for 19.6.2 and 19.5.4

### PR DESCRIPTION
**Pull request by proller**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,8 +1,75 @@
+## ClickHouse release 19.6.2.11, 2019-05-13
+
+### Bug Fix
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: Categories are not sorted.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,8 +1,75 @@
...
+* Partially fix distributed_product_mode=local. It's possible to allow columns of local tables in where/having/order by/... via table aliases. Throw exception if table does not have alias. There's not possible to access to the columns without table aliases yet. [#4986](https://github.com/yandex/ClickHouse/pull/4986) ([Artem Zuikov](https://github.com/4ertus2))
+* Function `toISOWeek` returned wrong results for year 1970. [#4988](https://github.com/yandex/ClickHouse/pull/4988) ([alexey-milovidov](https://github.com/alexey-milovidov))
+* Now follower replicas would not send queries to leader if they receive them from DDL. [#4991](https://github.com/yandex/ClickHouse/pull/4991) ([alesapin](https://github.com/alesapin))
+* Fix wrong result for 'select distinct' with multiple streams (ex. leaded by join) [#5001](https://github.com/yandex/ClickHouse/pull/5001) ([Artem Zuikov](https://github.com/4ertus2))
+* The `max_partitions_per_insert_block` field is of `size_t` type, rather than `Boolean` [#5028](https://github.com/yandex/ClickHouse/pull/5028) ([Mohammad Hossein Sekhavat](https://github.com/mhsekhavat))
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: That's not a description of a bug fix.

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,8 +1,75 @@
...
+* Set alias for external table (right part of GLOBAL JOIN) to be able to resolve its columns. [#4969](https://github.com/yandex/ClickHouse/pull/4969) ([Artem Zuikov](https://github.com/4ertus2))
+* Partially fix distributed_product_mode=local. It's possible to allow columns of local tables in where/having/order by/... via table aliases. Throw exception if table does not have alias. There's not possible to access to the columns without table aliases yet. [#4986](https://github.com/yandex/ClickHouse/pull/4986) ([Artem Zuikov](https://github.com/4ertus2))
+* Function `toISOWeek` returned wrong results for year 1970. [#4988](https://github.com/yandex/ClickHouse/pull/4988) ([alexey-milovidov](https://github.com/alexey-milovidov))
+* Now follower replicas would not send queries to leader if they receive them from DDL. [#4991](https://github.com/yandex/ClickHouse/pull/4991) ([alesapin](https://github.com/alesapin))
+* Fix wrong result for 'select distinct' with multiple streams (ex. leaded by join) [#5001](https://github.com/yandex/ClickHouse/pull/5001) ([Artem Zuikov](https://github.com/4ertus2))
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: Bad style.

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,8 +1,75 @@
+## ClickHouse release 19.6.2.11, 2019-05-13
+
+### Bug Fix
+* Add locks for ASOF lookup data cause it has multithreaded access. Replace some macro code with templates. [#4875](https://github.com/yandex/ClickHouse/pull/4875) ([Artem Zuikov](https://github.com/4ertus2))
+* Fixed potential null pointer dereference in `clickhouse-copier`. [#4900](https://github.com/yandex/ClickHouse/pull/4900) ([proller](https://github.com/proller))
+* Fix bug in codecs parameters detection in `ALTER MODIFY` of codec without type of column. [#4883](https://github.com/yandex/ClickHouse/pull/4883) ([alesapin](https://github.com/alesapin))
+* Functions cutQueryStringAndFragment() and queryStringAndFragment() now works correctly when url contains a fragment and no query. [#4894](https://github.com/yandex/ClickHouse/pull/4894) ([Vitaly Baranov](https://github.com/vitlibar))
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: Extra parentheses. This is not a `C` language.

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,8 +1,75 @@
...
+* Fix bug in `ReadBufferAIO` which was not able to seek backwards after EOF (now it can). In case of concurrent `SELECT` query some threads may "steal" portions of read-data from other threads and seek backwards in buffers. It leads to `Can't adjust last granule` error in rare cases if setting `min_bytes_to_use_direct_io` was set to positive number. [#4897](https://github.com/yandex/ClickHouse/pull/4897) ([alesapin](https://github.com/alesapin))
+* Fix bug with additional file seek and read when block already in uncompressed cached. [#4913](https://github.com/yandex/ClickHouse/pull/4913) ([alesapin](https://github.com/alesapin))
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: Typo.

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,8 +1,75 @@
...
+* Fixed wrong argument types for aggregate functions with low cardinality arguments. https://github.com/yandex/ClickHouse/issues/4919 [#4922](https://github.com/yandex/ClickHouse/pull/4922) ([Nikolai Kochetov](https://github.com/KochetovNicolai))
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: That may lead to ...?

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,8 +1,75 @@
...
+* If multiple joins are enabled there were an error on query with JOIN + ARRAY JOIN [#4938](https://github.com/yandex/ClickHouse/pull/4938) ([Artem Zuikov](https://github.com/4ertus2))
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: They are enabled by default.

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,8 +1,75 @@
...
+* Fix hanging on start of the server when a dictionary depends on another dictionary via a database with engine=Dictionary. [#4962](https://github.com/yandex/ClickHouse/pull/4962) ([Vitaly Baranov](https://github.com/vitlibar))
+* Set alias for external table (right part of GLOBAL JOIN) to be able to resolve its columns. [#4969](https://github.com/yandex/ClickHouse/pull/4969) ([Artem Zuikov](https://github.com/4ertus2))
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: What is the bug that was fixed?

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,8 +1,75 @@
...
+* Use another allocator for MarkCache and UncompressedCache to catch segfaults faster. I had to move functions to inline definition because otherwise I would need to copy-paste a lot of code, for now the code is fully reused but compilation time can increase a bit [#4928](https://github.com/yandex/ClickHouse/pull/4928) ([Danila Kutenin](https://github.com/danlark1))
+* Deduplicate pointers in ASTSelectQuery and its children cause they could became not consistent. Leave one copy + positions map. [#4952](https://github.com/yandex/ClickHouse/pull/4952) ([Artem Zuikov](https://github.com/4ertus2))
+* Added support for non-constant and negative size and length arguments for function 'substringUTF8'. [#4989](https://github.com/yandex/ClickHouse/pull/4989) ([alexey-milovidov](https://github.com/alexey-milovidov))
+* Yandex LFAlloc was added to ClickHouse to allocate MarkCache and UncompressedCache data in different ways to catch segfaults more reliable [#4995](https://github.com/yandex/ClickHouse/pull/4995) ([Danila Kutenin](https://github.com/danlark1))
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: This is a testing improvement.

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,8 +1,75 @@
...
+* Now block size in rows is bounded by fixed granularity. [#5052](https://github.com/yandex/ClickHouse/pull/5052) ([alesapin](https://github.com/alesapin))
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: No idea what does it mean.

File: CHANGELOG.md [hunk 11]
```diff
@@ -1,8 +1,75 @@
...
+
+### Other
+* Python util to help with backports and changelogs. [#4949](https://github.com/yandex/ClickHouse/pull/4949) ([Ivan](https://github.com/abyss7))
+
+### Backward Incompatible Change
+* HTTP header `Query-Id` was renamed to `X-ClickHouse-Query-Id` for consistency. [#4972](https://github.com/yandex/ClickHouse/pull/4972) ([Mikhail ](https://github.com/fandyushin))
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: Mikhail extra whitespace.

File: CHANGELOG.md [hunk 12]
```diff
@@ -1,8 +1,75 @@
...
+
+### Performance Improvement
+* Previously all of the AsofRowRefs were protected by a single mutex per Join object, which is obviously not ideal from a performance perspective as joinBlock becomes essentially single threaded (for the non-asof case this can be called by multiple threads simultaneously on their own blocks as they do not modify the state of the Join object). I moved the lock into the SortedLookupVector instead, which distributes them between the different keys. The reason for the need of the lock is that the AsofRowRefs only sorts it's contents when it's required (and therefore isn't strictly read-only, but is accessed in the JoinBlock stage of the Join). However, it is logically read-only, as the user should not observe any changes, merely the faster lookup. ... [#4924](https://github.com/yandex/ClickHouse/pull/4924) ([Martijn Bakker](https://github.com/Gladdy))
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: Not a description for the changelog.

File: CHANGELOG.md [hunk 13]
```diff
@@ -1,8 +1,75 @@
...
+* Use radixSort instead of std::sort for ASOF JOIN values. [#4993](https://github.com/yandex/ClickHouse/pull/4993) ([Artem Zuikov](https://github.com/4ertus2))
+
+### New Feature
+* Implements TTL for columns and tables. [#4212](https://github.com/yandex/ClickHouse/pull/4212) ([Anton Popov](https://github.com/CurtizJ))
+* Add setting `index_granularity_bytes` for MergeTree* tables family. ... [#4826](https://github.com/yandex/ClickHouse/pull/4826) ([alesapin](https://github.com/alesapin))
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: Experimental feature.

File: CHANGELOG.md [hunk 14]
```diff
@@ -1,8 +1,75 @@
...
+* Using brotli compression for HTTP responses (Accept-Encoding: br) [#4388](https://github.com/yandex/ClickHouse/pull/4388) ([Mikhail ](https://github.com/fandyushin))
+* Added a new aggregate function leastSqr(x, y) which performs linear regression on points (x, y) and returns the parameters of the line. (from #4668) [#4917](https://github.com/yandex/ClickHouse/pull/4917) ([Nikolai Kochetov](https://github.com/KochetovNicolai))
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: Wrong name?

File: CHANGELOG.md [hunk 15]
```diff
@@ -1,8 +1,75 @@
...
+* isValidUTF8 function added for validating if the set of bytes is correctly utf-8 encoded. ... [#4934](https://github.com/yandex/ClickHouse/pull/4934) ([Danila Kutenin](https://github.com/danlark1))
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: ...?

File: CHANGELOG.md [hunk 16]
```diff
@@ -1,8 +1,75 @@
...
+* Add new load balancing policy for picking replicas (first_or_random). Useful for cross-replication topology setups. [#5012](https://github.com/yandex/ClickHouse/pull/5012) ([nvartolomei](https://github.com/nvartolomei))
+
+
+## ClickHouse release 19.5.4.22, 2019-05-13
+
+### Bug fixes
+* Fixed possible crash in bitmap* functions #5220  [#5228](https://github.com/yandex/ClickHouse/pull/5228) ([Andy Yang](https://github.com/andyyzh))
+* Fixed very rare data race condition that could happen when executing a query with UNION ALL involving at least two SELECTs from system.columns, system.tables, system.parts, system.parts_tables or tables of Merge family and performing ALTER of columns of the related tables concurrently. [#5189](https://github.com/yandex/ClickHouse/pull/5189) ([alexey-milovidov](https://github.com/alexey-milovidov))
+* Fixed error Set for IN is not created yet in case of using single LowCardinality column in the left part of IN. This error happened if LowCardinality column was the part of primary key. #5031 [#5154](https://github.com/yandex/ClickHouse/pull/5154) ([Nikolai Kochetov](https://github.com/KochetovNicolai))
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: No quoting of the error message.

File: CHANGELOG.md [hunk 17]
```diff
@@ -1,8 +1,75 @@
...
+* Modification retention function: If a row satisfies both the first and NTH condition, only the first satisfied condition is added to the data state. Now all conditions that satisfy in a row of data are added to the data state. [#5119](https://github.com/yandex/ClickHouse/pull/5119) ([小路](https://github.com/nicelulu))
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: `Modification retention function` - illiterate

File: CHANGELOG.md [hunk 18]
```diff
@@ -1,8 +1,75 @@
...
+* Something bizarre happens if we have default move-constructor without default move-assignment: segfaults without stacktraces deep inside allocators.  [#5143](https://github.com/yandex/ClickHouse/pull/5143) ([Ivan](https://github.com/abyss7))
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: Actually not a bug. Irrelevant.

File: CHANGELOG.md [hunk 19]
```diff
@@ -1,8 +1,75 @@
...
+* Using brotli compression for HTTP responses (Accept-Encoding: br) [#4388](https://github.com/yandex/ClickHouse/pull/4388) ([Mikhail ](https://github.com/fandyushin))
+* Added a new aggregate function leastSqr(x, y) which performs linear regression on points (x, y) and returns the parameters of the line. (from #4668) [#4917](https://github.com/yandex/ClickHouse/pull/4917) ([Nikolai Kochetov](https://github.com/KochetovNicolai))
+* isValidUTF8 function added for validating if the set of bytes is correctly utf-8 encoded. ... [#4934](https://github.com/yandex/ClickHouse/pull/4934) ([Danila Kutenin](https://github.com/danlark1))
+* Add new load balancing policy for picking replicas (first_or_random). Useful for cross-replication topology setups. [#5012](https://github.com/yandex/ClickHouse/pull/5012) ([nvartolomei](https://github.com/nvartolomei))
+
+
+## ClickHouse release 19.5.4.22, 2019-05-13
+
+### Bug fixes
+* Fixed possible crash in bitmap* functions #5220  [#5228](https://github.com/yandex/ClickHouse/pull/5228) ([Andy Yang](https://github.com/andyyzh))
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: Missing link to #5220.
Double whitespace.

File: CHANGELOG.md [hunk 20]
```diff
@@ -1,8 +1,75 @@
...
+
+### New Feature
+* Implements TTL for columns and tables. [#4212](https://github.com/yandex/ClickHouse/pull/4212) ([Anton Popov](https://github.com/CurtizJ))
+* Add setting `index_granularity_bytes` for MergeTree* tables family. ... [#4826](https://github.com/yandex/ClickHouse/pull/4826) ([alesapin](https://github.com/alesapin))
+* Using brotli compression for HTTP responses (Accept-Encoding: br) [#4388](https://github.com/yandex/ClickHouse/pull/4388) ([Mikhail ](https://github.com/fandyushin))
+* Added a new aggregate function leastSqr(x, y) which performs linear regression on points (x, y) and returns the parameters of the line. (from #4668) [#4917](https://github.com/yandex/ClickHouse/pull/4917) ([Nikolai Kochetov](https://github.com/KochetovNicolai))
+* isValidUTF8 function added for validating if the set of bytes is correctly utf-8 encoded. ... [#4934](https://github.com/yandex/ClickHouse/pull/4934) ([Danila Kutenin](https://github.com/danlark1))
+* Add new load balancing policy for picking replicas (first_or_random). Useful for cross-replication topology setups. [#5012](https://github.com/yandex/ClickHouse/pull/5012) ([nvartolomei](https://github.com/nvartolomei))
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: What does it mean actually?

File: CHANGELOG.md [hunk 21]
```diff
@@ -1,8 +1,75 @@
...
+* Use radixSort instead of std::sort for ASOF JOIN values. [#4993](https://github.com/yandex/ClickHouse/pull/4993) ([Artem Zuikov](https://github.com/4ertus2))
+
+### New Feature
+* Implements TTL for columns and tables. [#4212](https://github.com/yandex/ClickHouse/pull/4212) ([Anton Popov](https://github.com/CurtizJ))
+* Add setting `index_granularity_bytes` for MergeTree* tables family. ... [#4826](https://github.com/yandex/ClickHouse/pull/4826) ([alesapin](https://github.com/alesapin))
+* Using brotli compression for HTTP responses (Accept-Encoding: br) [#4388](https://github.com/yandex/ClickHouse/pull/4388) ([Mikhail ](https://github.com/fandyushin))
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: Added support for `brotli` compression,
not "using brotli compression".

File: CHANGELOG.md [hunk 22]
```diff
@@ -1,8 +1,75 @@
...
+* Copier: auto upload task configuration from --task-file option [#4876](https://github.com/yandex/ClickHouse/pull/4876) ([proller](https://github.com/proller))
+* Added typos handler for storage factory and table functions factory. [#4891](https://github.com/yandex/ClickHouse/pull/4891) ([Danila Kutenin](https://github.com/danlark1))
+* Support asterisks and qualified asterisks for multiple joins without subqueries [#4898](https://github.com/yandex/ClickHouse/pull/4898) ([Artem Zuikov](https://github.com/4ertus2))
+* Make missing column error message more user friendly. [#4915](https://github.com/yandex/ClickHouse/pull/4915) ([Artem Zuikov](https://github.com/4ertus2))
+* Use another allocator for MarkCache and UncompressedCache to catch segfaults faster. I had to move functions to inline definition because otherwise I would need to copy-paste a lot of code, for now the code is fully reused but compilation time can increase a bit [#4928](https://github.com/yandex/ClickHouse/pull/4928) ([Danila Kutenin](https://github.com/danlark1))
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: Duplicate.

File: CHANGELOG.md [hunk 23]
```diff
@@ -1,8 +1,75 @@
...
+* Deduplicate pointers in ASTSelectQuery and its children cause they could became not consistent. Leave one copy + positions map. [#4952](https://github.com/yandex/ClickHouse/pull/4952) ([Artem Zuikov](https://github.com/4ertus2))
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: What bug was fixed?

File: CHANGELOG.md [hunk 24]
```diff
@@ -1,8 +1,75 @@
...
+* Function `toISOWeek` returned wrong results for year 1970. [#4988](https://github.com/yandex/ClickHouse/pull/4988) ([alexey-milovidov](https://github.com/alexey-milovidov))
+* Now follower replicas would not send queries to leader if they receive them from DDL. [#4991](https://github.com/yandex/ClickHouse/pull/4991) ([alesapin](https://github.com/alesapin))
+* Fix wrong result for 'select distinct' with multiple streams (ex. leaded by join) [#5001](https://github.com/yandex/ClickHouse/pull/5001) ([Artem Zuikov](https://github.com/4ertus2))
+* The `max_partitions_per_insert_block` field is of `size_t` type, rather than `Boolean` [#5028](https://github.com/yandex/ClickHouse/pull/5028) ([Mohammad Hossein Sekhavat](https://github.com/mhsekhavat))
+
+### Build/Testing/Packaging Improvement
+* Further improvement of the docker-based test machinery ... [#4713](https://github.com/yandex/ClickHouse/pull/4713) ([Vasily Nemkov](https://github.com/Enmk))
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: ...?

File: CHANGELOG.md [hunk 25]
```diff
@@ -1,8 +1,75 @@
...
+*  [#4937](https://github.com/yandex/ClickHouse/pull/4937) ([alesapin](https://github.com/alesapin))
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: ?

File: CHANGELOG.md [hunk 26]
```diff
@@ -1,8 +1,75 @@
...
+* Tests: Allow use any test database (remove test.hardcode where it possible) [#5008](https://github.com/yandex/ClickHouse/pull/5008) ([proller](https://github.com/proller))
+* Fix ubsan errors [#5037](https://github.com/yandex/ClickHouse/pull/5037) ([Vitaly Baranov](https://github.com/vitlibar))
+
+### Improvement
+* Disable push-down to right table in left join, left table in right join, and both tables in full join. [#4846](https://github.com/yandex/ClickHouse/pull/4846) ([Ivan](https://github.com/abyss7))
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: No explanation why it is an improvement.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
.

File: CHANGELOG.md (hunk 26 above)
Comment (vitlibar) [reply CHANGELOG.md]: @abyss7

File: CHANGELOG.md [hunk 27]
```diff
@@ -1,8 +1,66 @@
+## ClickHouse release 19.6.2.11, 2019-05-13
+
+### Backward Incompatible Change
+* HTTP header `Query-Id` was renamed to `X-ClickHouse-Query-Id` for consistency. [#4972](https://github.com/yandex/ClickHouse/pull/4972) ([Mikhail](https://github.com/fandyushin))
+
+### New Feature
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: Bad title.

File: CHANGELOG.md [hunk 28]
```diff
@@ -1,8 +1,66 @@
+## ClickHouse release 19.6.2.11, 2019-05-13
+
+### Backward Incompatible Change
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: Wrong order.

File: CHANGELOG.md [hunk 29]
```diff
@@ -1,8 +1,66 @@
...
+* Added support for `brotli` compression for HTTP responses (Accept-Encoding: br) [#4388](https://github.com/yandex/ClickHouse/pull/4388) ([Mikhail](https://github.com/fandyushin))
+* Added a new aggregate function leastSqr(x, y) which performs linear regression on points (x, y) and returns the parameters of the line. (from #4668) [#4917](https://github.com/yandex/ClickHouse/pull/4917) ([hcz](https://github.com/hczhcz))
+* Added new function isValidUTF8 for checking whether a set of bytes is correctly utf-8 encoded. [#4934](https://github.com/yandex/ClickHouse/pull/4934) ([Danila Kutenin](https://github.com/danlark1))
+* Add new load balancing policy for picking replicas (first_or_random). Useful for cross-replication topology setups. [#5012](https://github.com/yandex/ClickHouse/pull/5012) ([nvartolomei](https://github.com/nvartolomei))
+
+### Experimental features
+* Add setting `index_granularity_bytes` for MergeTree* tables family. ... [#4826](https://github.com/yandex/ClickHouse/pull/4826) ([alesapin](https://github.com/alesapin))
+
+### Bug Fix
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: Wrong title.

File: CHANGELOG.md [hunk 30]
```diff
@@ -1,8 +1,66 @@
...
+* Fixed potential null pointer dereference in `clickhouse-copier`. [#4900](https://github.com/yandex/ClickHouse/pull/4900) ([proller](https://github.com/proller))
+* Fixed error on query with JOIN + ARRAY JOIN [#4938](https://github.com/yandex/ClickHouse/pull/4938) ([Artem Zuikov](https://github.com/4ertus2))
+* Fixed hanging on start of the server when a dictionary depends on another dictionary via a database with engine=Dictionary. [#4962](https://github.com/yandex/ClickHouse/pull/4962) ([Vitaly Baranov](https://github.com/vitlibar))
+* Partially fix distributed_product_mode=local. It's possible to allow columns of local tables in where/having/order by/... via table aliases. Throw exception if table does not have alias. There's not possible to access to the columns without table aliases yet. [#4986](https://github.com/yandex/ClickHouse/pull/4986) ([Artem Zuikov](https://github.com/4ertus2))
+* Fixed wrong result for 'select distinct' with multiple streams (ex. leaded by join) [#5001](https://github.com/yandex/ClickHouse/pull/5001) ([Artem Zuikov](https://github.com/4ertus2))
+
+### Performance Improvement
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: Wrong title.

File: CHANGELOG.md [hunk 31]
```diff
@@ -1,8 +1,66 @@
...
+* Significant speedup of ASOF join [#4924](https://github.com/yandex/ClickHouse/pull/4924) ([Martijn Bakker](https://github.com/Gladdy))
+* Use radixSort instead of std::sort for ASOF JOIN values. [#4993](https://github.com/yandex/ClickHouse/pull/4993) ([Artem Zuikov](https://github.com/4ertus2))
+
+### Improvement
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: Wrong title.

File: CHANGELOG.md [hunk 32]
```diff
@@ -1,8 +1,66 @@
...
+* Make missing column error message more user friendly. [#4915](https://github.com/yandex/ClickHouse/pull/4915) ([Artem Zuikov](https://github.com/4ertus2))
+* Use another allocator for MarkCache and UncompressedCache to catch segfaults faster. [#4928](https://github.com/yandex/ClickHouse/pull/4928) ([Danila Kutenin](https://github.com/danlark1))
+* Deduplicate pointers in ASTSelectQuery and its children cause they could became not consistent. Leave one copy + positions map. [#4952](https://github.com/yandex/ClickHouse/pull/4952) ([Artem Zuikov](https://github.com/4ertus2))
+* Added support for non-constant and negative size and length arguments for function 'substringUTF8'. [#4989](https://github.com/yandex/ClickHouse/pull/4989) ([alexey-milovidov](https://github.com/alexey-milovidov))
+* Use fixed_granularity as upper bound for adaptive granularity. Now block size in rows is bounded by fixed granularity. [#5052](https://github.com/yandex/ClickHouse/pull/5052) ([alesapin](https://github.com/alesapin))
+
+### Build/Testing/Packaging Improvement
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: Wrong title.

File: CHANGELOG.md [hunk 33]
```diff
@@ -1,8 +1,66 @@
...
+* Fixed test failures when running clickhouse-server on different host [#4713](https://github.com/yandex/ClickHouse/pull/4713) ([Vasily Nemkov](https://github.com/Enmk))
+* Tests: Disable color control sequences in non tty environment. [#4937](https://github.com/yandex/ClickHouse/pull/4937) ([alesapin](https://github.com/alesapin))
+* Tests: Allow use any test database (remove test.hardcode where it possible) [#5008](https://github.com/yandex/ClickHouse/pull/5008) ([proller](https://github.com/proller))
+* Fix ubsan errors [#5037](https://github.com/yandex/ClickHouse/pull/5037) ([Vitaly Baranov](https://github.com/vitlibar))
+* Yandex LFAlloc was added to ClickHouse to allocate MarkCache and UncompressedCache data in different ways to catch segfaults more reliable [#4995](https://github.com/yandex/ClickHouse/pull/4995) ([Danila Kutenin](https://github.com/danlark1))
+
+### Other
```
Comment (alexey-milovidov) [inline CHANGELOG.md]: This is build build/testing/packaging improvement.

File: CHANGELOG.md (hunk 28 above)
Comment (vitlibar) [reply CHANGELOG.md]: What order is correct?

File: CHANGELOG.md (hunk 20 above)
Comment (vitlibar) [reply CHANGELOG.md]: @alesapin

File: CHANGELOG.md (hunk 3 above)
Comment (vitlibar) [reply CHANGELOG.md]: @4ertus2

File: CHANGELOG.md (hunk 23 above)
Comment (vitlibar) [reply CHANGELOG.md]: @4ertus2

File: CHANGELOG.md (hunk 3 above)
Comment (vitlibar) [reply CHANGELOG.md]: What's wrong with the style?

File: CHANGELOG.md (hunk 23 above)
Comment (4ertus2) [reply CHANGELOG.md]: The only known bug with #4952 was error in constants optimisation described here https://github.com/yandex/ClickHouse/issues/4242#issuecomment-477250658
But there could be some unknown ones cause AST was inconsistent after optimisation.

File: CHANGELOG.md (hunk 3 above)
Comment (4ertus2) [reply CHANGELOG.md]: Maybe 'fix wrong result for select distinct with join'

File: CHANGELOG.md (hunk 32 above)
Comment (vitlibar) [reply CHANGELOG.md]: Renamed to Build/Testing/Packaging Improvement**s**

File: CHANGELOG.md (hunk 31 above)
Comment (vitlibar) [reply CHANGELOG.md]: Renamed to improvement**s**

File: CHANGELOG.md (hunk 30 above)
Comment (vitlibar) [reply CHANGELOG.md]: Renamed to Performance Improvement**s**

File: CHANGELOG.md (hunk 29 above)
Comment (vitlibar) [reply CHANGELOG.md]: Renamed to Bug Fix**es**

File: CHANGELOG.md (hunk 28 above)
Comment (vitlibar) [reply CHANGELOG.md]: Reordered

File: CHANGELOG.md (hunk 27 above)
Comment (vitlibar) [reply CHANGELOG.md]: Renamed to New Feature**s**

**Pull request merged**